From 521080bbbc721845a1c681a3f01e6f835283346e Mon Sep 17 00:00:00 2001
From: Yixin Shou <yixin.shou@intel.com>
Date: Fri, 5 Apr 2013 11:03:35 -0700
Subject: Dalvik: fix registerization syncCompileTableWithAssociations

BZ: 98245

At start of basic block, compile table is updated based on the
association table from the parent using syncCompileTableWithAssociations.
compileTable entries might be overwritten wrongly by only comparing
the virtual register number. This patch fixed this bug by adding comparing
virtual register type.

Category: device-enablement
Domain: AOSP-Dalvik-Compiler-CG
Origin: internal
Upstream-Candidate: no, needs rework

Change-Id: I89936394287e57e70eeb94fcdf6afc40668a0a80
Orig-MCG-Change-Id: I86a7ee12d60889797c7ea0aa428166b21e4a2b1f
Signed-off-by: Yixin Shou <yixin.shou@intel.com>
Signed-off-by: Qiming Shi <qiming.shi@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/compiler/codegen/x86/RegisterizationBE.cpp |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/vm/compiler/codegen/x86/RegisterizationBE.cpp b/vm/compiler/codegen/x86/RegisterizationBE.cpp
index c017492..58a2c5b 100644
--- a/vm/compiler/codegen/x86/RegisterizationBE.cpp
+++ b/vm/compiler/codegen/x86/RegisterizationBE.cpp
@@ -328,6 +328,7 @@ bool AssociationTable::syncCompileTableWithAssociations(AssociationTable & assoc
 
             //If it is a virtual register and the right register
             if (isVirtualReg(compileTable[entry].physicalType) == true
+                    && compileTable[entry].physicalType == assocIter->second.physicalType
                     && compileTable[entry].regNum == assocIter->first) {
                 DEBUG_COMPILETABLE_UPDATE(ALOGD("Found that v%d is in compile "
                         "table already.", assocIter->first));
-- 
1.7.4.1

